design/one-pager-resource-connectivity-mvp.md

```diff
@@ -638,7 +638,279 @@ parameters:
+
+* `AwsSecurityGroup`: allows the cluster to communicate with worker nodes. It logically groups the resources that could communicate with each other within a VPC.
+* `AwsIamRole`: enables EKS to make calls to other AWS services to manage the resources.
+* `AwsIamRolePolicyAttachment`: attaches required policies the EKS role
+* `AwsInternetGateway`: enables the nodes to have traffic to and from the internet. This is necessary because most workloads have a UI that needs to be accessed from the internet.
+* `AwsRouteTable`: for routing internet traffic from `AwsSubnet`s to `AwsInternetGateway`
+* `AwsRoutTableAssociation`: associates an `AwsRouteTable` to a `AwsSubnet`
+
+In addition, `RDSInstance`s also need following resources, so that they are accessible by the the worker nodes:
+
+* `AwsDbSubnetGroup`: represents a group of `AwsSubnet`s from different availability zones, from which a private IP is chosen and assigned to the RDS instance. 
+* `AwsSecurityGroup`: allows the RDS instance to accept traffic from a certain IP and port.
+
+To achieve resource connectivity goal in AWS, two approaches can be imagined:
+
+1. **Configuration based single managed controller:** Encapsulate all the external resource types in a general `AwsNetwork` type and create a single managed resource for it.
+2. **External resource based managed controller**: Create a corresponding managed resource for each required external resource type
+
+Although the first approach requires potentially less effort, the latter approach provides a few advantages:
+
+1. **Reusability**: these managed resource types could be later re-used for other configuration
+1. **YAML Configuration**: instead of implementing a specific configuration logic inside `AwsNetwork`, the configuration details are expressed in a YAML file. This makes creating more sophisticated configurations possible, without having to write controllers. 
+
+This document focuses on implementing the second approach.  For doing so, we address the following challenges:
+
+* As pointed out [earlier](#resource-references) in this document, each managed resource can refer to other resources either by their external names, or their object reference. In AWS case, since we are taking a YAML configuration approach, interconnecting resource by api objects makes a lot of sense. At the same time it is still useful to re-use existing external resources. Because of these requirements, we introduce the notion of *reference type* as following:
+
+	```yaml
+	...
+	spec:
+	  ...
+	  someParameter:
+	    # one of {external, apiObject} values
+	    referenceType: apiObject
+      # a string of attributeName.namespacedName format (apiObject case), or arbitrary (external case)
+	    value: attributeNAme.sampleObject.sampleNamespace
```

Comment:
This is a super interesting idea to be able to essentially have multiple types for a field. I am not sure how this scales long-term yet (still thinking through it), but a few things that immediately come to mind:
* If implemented, this should exist in `crossplane-runtime` as our own form of a Kubernetes `ObjectReference`
* If implemented, we are essentially deferring logic to Crossplane controllers as opposed to core Kubernetes functionality. We should probably be conservative about packing more and more logic intro controllers. However, there may be an opportunity to include some of this generic functionality is the shared reconcilers.

Reply:
Agreed. I would suggest we start it as a controller scoped concept and see if we like it. We could then factored it out to runtime library.

Reply:
I had also considered an approach like this, but was wary that it lowers the fidelity our our representation of the cloud provider API. i.e. Instead of 'the value of every field is exactly what you'd submit to the AWS API', the contract becomes 'the value of every field is exactly what you'd submit to the AWS API, unless it's a reference'.

I recall someone (perhaps @bassam?) commenting that some kind of templating above the YAML level (e.g. Helm) could be more appropriate than teaching Crossplane how to do this (i.e. use a template variable to name the referenced object `foo`, and use the same variable to reference `foo` from object `bar`). I like this idea, but it falls short when it comes to referencing API objects whose names are not known until after they're created.

My preference is that we use purely 'external name' style references at first, then investigate adding something like what you've proposed here in future.

Reply:
I see your point how this could limiting, in terms of parameter representation. However, there are two reasons why purely using external resources will be challenging:
- As it's been pointed out in the document, interconnecting resources for a given configuration is done in the YAML file level. For instance you manifest the resource `foo` and `bar` in the YAML, where `bar` is consuming `foo`'s id:
```yaml
metadata:
  name: foo
  namespace: example
spec:
  ...
---
metadata:
  name: bar
  namespace: example
spec:
  ...
  fooId:
    referenceType: apiObject
    value: id.foo.example
  ...
```
Using api object references enables to keep coherency and logical connection of these two resources in the YAML description, whereas using the external names of `foo`'s id (e.g. `rnd87934575`) will be less meaningful.
 
- Perhaps the bigger challenge is the indeterministic nature of external resource's attributes (as you pointed out). For instance in AWS case, most resources's attribute values are not deterministic and only become available after provisioning. Now, if an administrator want to create a, say, resource connectivity configuration, they won't be able to apply a single YAML object. Instead they will need to create each external resource individually, and then inject the newly created resources attributes into the consuming resources. This will result in using a lot of potentially fragile scripting code (as is the case today) to do this parameter injection. 

Regarding the fidelity of the external resources, I think object references doesn't limit that, just changes the representation. Resource managers like `Terraform` and `Pulumi` already use domain specific aliases for corss referencing external resources.

Reply:
@soorena776 @negz @hasheddan @jbw976 met today to sync up on this inter-resource reference topic.  Below is a summary of the conclusion of that discussion:

* the " value resolver" approach is pretty compelling and a novel ability in the ecosystem that would enhance the automation and experience of provisioning multiple resources that  form a chain of dependencies (values from completed upstream resource are used as input to next downstream resource in the chain)
* Sufficient time will be needed to think through the experience and design of such a feature, and likely there is not enough time left in v0.3 to give it the attention it needs to really get it right.
* We will start very simple in v0.3 with simple string fields that just take the name of the external resource.  Values will need to be looked up by the user from a completed resource in order for the user to input them on the next resource.
* This is not an ideal user experience, but we can invest in some simple scripting and commands that the user can copy from the docs to help make this process easier and more reliable.
* We hope to finish coding and manual testing for resource connectivity CRDs and controllers by the end of this sprint and then we can spend the final sprint in v0.3 on docs, user guides, scripts, integration testing, and polishing.
* @soorena776 will take the write-up and thinking he's already done in this PR for the "value resolver" approach and open a new issue for that with the "proposals" label. We should consider taking that issue on early in v0.4 since it is such a user experience win.

Please correct anything if I captured it incorrectly, great discussion today team! 🚀

Reply:
Per offline discussion with @jbw976 @negz and @hasheddan, addressing inter-resource references separately in #707